### DISCUSSION TITLE
Accepts promise parameter instantiation

### DISCUSSION THREAD

**Discussion by lush-99** [sequelize/sequelize#14542] [Help & Questions]
Database information often needs to be encrypted，and We have the corresponding decryption algorithm，but this function return promise, Could you please provide the solution。

The best way is to instantiate the interface promise。

example：Constructor function can accept promis

**Comment by lush-99** [comment]
this func is this method is to accept the ciphertext and then output the password
`Async function decrypt ( ){
let password = await  func()
return password 
}`
i thinnk that is goods if i can 
`Sequelize sequelize = new Sequelize(username,decrypt(ciphertext),url)`

**Reply by ephys** [reply]
I need more information to be able to help you. What is encrypted? Is it the contents of a single column? All columns in all tables?
Is it the credentials to connect to the database?

**Reply by lush-99** [reply]
It is very insecure to write the database password directly in the online code. For the sake of security, we need to encrypt the database password. In this way, we only need to write the encrypted ciphertext in the code, and then use it when we need to link to the database. thrift, decrypt the ciphertext in the code to get the database password, and then link the database

**Reply by ephys** [reply]
You could consider using environment variables for that.

But if you want to go the encryption route, you'll need to delay creating the sequelize constructor until after you have decrypted your credentials:

```typescript
async function createSequelize(): Promise<Sequelize> {
  return new Sequelize(username, await decrypt(ciphertext), url);
}
```

If you're using ESM, [top-level await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await#top_level_await) can make that easy.

**Comment by lush-99** [comment]
i try it ;
but it have a troubled。

`
async function createSequelize(): Promise<Sequelize> {
  const sequelize = new Sequelize(username, await decrypt(ciphertext), url);
}
db.sequelize = sequelize;
var db = {};
module.exports = db;`

when i use :
`var sequelize = models.sequelize;`

this sequelize is undefined or null

**Comment by lush-99** [comment]
My current code is:
in the index file：
`var sequelize = new Sequelize(username, password,url）;
db.sequelize = sequelize;
var db = {};
module.exports = db;
`

in other documents:
`
import models = require('/index')
var sequelize = models.sequelize;
sequelize.query(xxxxxxx)
`
How do I use this asynchronous decryption method, I think the best way is to accept the promise in the constructor

**Reply by ephys** [reply]
We won't accept a promise in our constructor, we'd have to make our constructor async which is a no-go. It's also bad practice, it's on the caller to provide a resolved value, not on the callee to await all of its inputs.

In your code, Sequelize is undefined because you're never assigning a value to it.
Unfortunately I can't help much more, you'll need to learn more about how async stuff works in JavaScript for this. My best recommendation would be to use EcmaScript modules instead of your current CommonJS modules, which will let you use top-level await, and make this much easier to write

**Reply by lush-99** [reply]
ok, i see, thanks.

**Comment by lush-99** [comment]
i find a hook that is 'beforeConnect', but i current version is '^3.4.1', i don't know if v5 is compatible；

**Comment by lush-99** [comment]
where can i find the change from v3 to v5?

**Reply by lush-99** [reply]
Can you guarantee that version upgrades are fully backward compatible

**Reply by ephys** [reply]
They are not, that's why they are major releases

v3 to v4: https://sequelize.org/v4/manual/tutorial/upgrade-to-v4.html
v4 to v5: https://sequelize.org/v5/manual/upgrade-to-v5.html
v5 to v6: https://sequelize.org/docs/v6/other-topics/upgrade/

I don't think `beforeConnect` is going to help with this specific problem though

**Reply by lush-99** [reply]
I find that I have a lot to change if I want to use the hook

could you please give me a special version, I expect this version to use the beforeConnect hook only on the basis of ^3.4.1

**Reply by lush-99** [reply]
I have tried it on the v5 version, this beforeConnect hooks can solve the problem

**Reply by WikiRik** [reply]
We'll not work on v3-5 anymore. Only v6 and v7 (alpha) are currently supported
